### PR TITLE
feat: link with ethereum on nakama

### PR DESCRIPTION
**Pull request by zulkhair**
Closes: GDEV-1106

## Overview

Add validation for SIWE in LinkCustom on nakama.

## Brief Changelog

- Add validation for SIWE before triggering LinkCustom

## Testing and Verifying

- Covered by old unit test

Manually tested using nakama console
- Before link

- Link Request

- After Link

### CODE REVIEW

**Comment by zulkhair** [comment]
* **#778** <a href="https://app.graphite.dev/github/pr/Argus-Labs/world-engine/778?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @zulkhair and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: relay/nakama/auth/siwe.go [hunk 1]
```diff
@@ -34,48 +34,87 @@ func validateSIWE(signer, message, signature string) (isValidationRequest bool,
...
-		return nil, err
+		return false, err
 	}
+
 	// The user has successfully been authenticated
-	return in, nil
+	return true, nil
+}
+
+func handleSIWEforAuth(
```
Comment (smsunarto) [inline relay/nakama/auth/siwe.go:78]: nit: rename this to `AuthWithSIWE`

File: relay/nakama/auth/siwe.go [hunk 2]
```diff
@@ -34,48 +34,87 @@ func validateSIWE(signer, message, signature string) (isValidationRequest bool,
...
+	}
+
+	if success {
+		return in, nil
+	}
+
+	return nil, err
+}
+
+func handleSIWEforLink(
```
Comment (smsunarto) [inline relay/nakama/auth/siwe.go:100]: nit: rename this to `LinkWithSIWE`

File: relay/nakama/auth/siwe.go [hunk 3]
```diff
@@ -34,48 +34,87 @@ func validateSIWE(signer, message, signature string) (isValidationRequest bool,
...
-	in *api.AuthenticateCustomRequest,
-) (
-	*api.AuthenticateCustomRequest, error,
-) {
-	signerAddress := in.GetAccount().GetId()
-	signature := in.GetAccount().GetVars()["signature"]
-	message := in.GetAccount().GetVars()["message"]
-
+	signerAddress, signature, message string,
+) (bool, error) {
```
Comment (smsunarto) [inline relay/nakama/auth/siwe.go:42]: You don't need to return a boolean, you can see whether or not the processing is successful by just checking whether or not err is nil.

File: relay/nakama/auth/siwe.go [hunk 4]
```diff
@@ -34,48 +34,87 @@ func validateSIWE(signer, message, signature string) (isValidationRequest bool,
...
+	success, err := processSIWE(ctx, logger, nk, signerAddress, signature, message)
+	if err != nil {
+		return nil, err
+	}
+
+	if success {
+		return in, nil
+	}
+
+	return nil, err
```
Comment (smsunarto) [inline relay/nakama/auth/siwe.go:119]: For readability, it's typically encouraged to have the last return statement be the "success" case.

**Review by smsunarto — CHANGES_REQUESTED** [review]
left some readability nits

File: relay/nakama/auth/siwe.go (hunk 4 above)
Comment (zulkhair) [reply relay/nakama/auth/siwe.go]: updated

File: relay/nakama/auth/siwe.go (hunk 3 above)
Comment (zulkhair) [reply relay/nakama/auth/siwe.go]: updated

File: relay/nakama/auth/siwe.go (hunk 2 above)
Comment (zulkhair) [reply relay/nakama/auth/siwe.go]: updated

File: relay/nakama/auth/siwe.go (hunk 1 above)
Comment (zulkhair) [reply relay/nakama/auth/siwe.go]: updated

**Review by smsunarto — APPROVED** [review]

**Pull request merged**